### PR TITLE
pinmux: serdes muxing support

### PR DESCRIPTION
**Pull request by merbanan**
Untested WIP suggestion on how to add serdes muxing support.

The same structure could also be used to re-add led mapping also.

If the approach is fine I will complete the implementation.

### CODE REVIEW

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 1]
```diff
@@ -218,19 +218,49 @@
 #define GPIO17_FLASH_MODE_CFG			BIT(1)
 #define GPIO16_FLASH_MODE_CFG			BIT(0)
 
+/* System Setting Register 3 */
+#define REG_SSR3				0x00
+#define SSUSB_HSGMII_SEL			BIT(29)
+#define ETH_XSI_SEL_MASK			GENMASK(14, 13)
+#define ETH_XSI_XFI_USXGMII			(1 << 13)
+#define ETH_XSI_SGMII_HSGMII			(2 << 13)
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan please use BIT() and GENMASK() macro here

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 2]
```diff
@@ -218,19 +218,49 @@
...
+#define PON_XSI_SEL_MASK			GENMASK(10, 9)
+#define PON_XSI1_XPON				(0 << 9)
+#define PON_XSI1_XFI_USXGMII			(1 << 9)
+#define PON_XSI1_SGMII_HSGMII			(2 << 9)
+#define USB_PCIE_SEL				BIT(3)
+#define USB_PCIE_P2_MODE			(0 << 3)
+#define USB_PCIE_USB_P1_MODE			(1 << 3)
+
+
+
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c:247]: @merbanan same as above + remove unnecessary new-lines

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 3]
```diff
@@ -1318,12 +1436,107 @@ static const struct airoha_pinctrl_func_group pwm_func_group[] = {
...
+	{
+		.name = "usb_hsgmii",
+		.mux_func = AIROHA_FUNC_SERDES_SSR3,
+		.regs[0] = {
+			REG_SSR3,
+			SSUSB_HSGMII_SEL,
+			SSUSB_HSGMII_SEL
+		},
+		.num_regs = 1,
+	},
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan no new-line here (and for all other cases below)

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 4]
```diff
@@ -1681,7 +1895,8 @@ static int airoha_pinmux_set_mux(struct pinctrl_dev *pctrl_dev,
 		for (j = 0; j < group->num_regs; j++)
 			airoha_pinctrl_mux_set(pinctrl, base,
 					       group->regs[j].offset,
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan I think here we should just use `airoha_pinctrl_rmw()`

**Comment by LorenzoBianconi** [comment]
@merbanan the code seems fine to me (just few cosmetics changes). I guess we should test it before merge it (in orther to avoid future rework)

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 3 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 4 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 1 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack

**Comment by merbanan** [comment]
Addressed all comments, dts changes missing. Compile tested only.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 5]
```diff
@@ -1604,8 +1854,8 @@ static u32 airoha_pinctrl_rmw(struct airoha_pinctrl *pinctrl,
 	return val;
 }
 
-#define airoha_pinctrl_mux_set(pinctrl, base, offset, val)			\
-	airoha_pinctrl_rmw((pinctrl), (base) + (offset), 0, (val));
+#define airoha_pinctrl_mux_set(pinctrl, base, offset, mask, val)			\
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan we do not need this change anymore

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 6]
```diff
@@ -329,6 +359,31 @@ static struct pinctrl_pin_desc airoha_pinctrl_pins[] = {
 	PINCTRL_PIN(61, "pcie_reset0"),
 	PINCTRL_PIN(62, "pcie_reset1"),
 	PINCTRL_PIN(63, "pcie_reset2"),
+	// SERDES PINS
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: @merbanan no C++ style comment

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c [hunk 7]
```diff
@@ -256,7 +286,7 @@ struct airoha_pinctrl {
 
 	struct mutex mutex;
 	struct {
-		void __iomem *mux[3];
+		void __iomem *mux[5];
```
Comment (LorenzoBianconi) [inline target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c:289]: @merbanan do we really need 2 separated mapping for serdes? do you have any dts example?

**Comment by LorenzoBianconi** [comment]
@merbanan I am fine with the code itself, but I guess we should test it before merging in order to avoid possible future changes. What do you think?

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 7 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c:289]: 1 register in the middle.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 6 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack.

File: target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c (hunk 5 above)
Comment (merbanan) [reply target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c]: Ack.

**Comment by merbanan** [comment]
> @merbanan I am fine with the code itself, but I guess we should test it before merging in order to avoid possible future changes. What do you think?

Yes, so far it is only compile tested. I will update when I have actually tested on real hardware.

**Comment by LorenzoBianconi** [comment]
@merbanan is this PR still needed?

**Comment by merbanan** [comment]
Well if you want to support any other serdes configuration but the defaults you need some functionality like this patch provides. As example I think that the ref board you have will need to change the eth xsi from xgi/usxgmii to sgmii/hsgmii for the phy/mac link to work. In some cases we might want to change the phy/mac link-mode in runtime. And the mode change register is shared between several macs so I dont see any better option then to add it to the pinmux framework.

**Pull request closed without merging**